### PR TITLE
feature: smdataparallel enable EFA RDMA flag

### PR DESCRIPTION
**Pull request by karan6181**
*Description of changes:*
- Enabled the EFA's RDMA functionality for one-sided and two-sided transfer for `ml.p4d.24xlarge` instances.

*Testing done:*
- Ran the smdataparallel unit test
- Ran the TF2+MNIST smdataparallel jupyter notebook on `p3.16xlarge` and `p4d.24xlarge`, both ran successfully.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have used the regional endpoint when creating S3 and/or STS clients (if appropriate)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-training-toolkit/blob/master/README.md)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by karan6181** [comment]
- Ran the TF2+MNIST smdataparallel jupyter notebook on `p3.16xlarge` and `p3dn.24xlarge`, both ran successfully.
- For `p3dn.24xlarge` instance, it adds `-x FI_EFA_USE_DEVICE_RDMA=1` to the mpirun command and for `p3.16xlarge` instance, it's not, which is expected.
- Attached the jupyter notebook zip file for the same
[smdataparallel_TF2_mnist_rdma_flag.zip](https://github.com/aws/sagemaker-training-toolkit/files/6266527/smdataparallel_TF2_mnist_rdma_flag.zip)

**Review by piyushghai — DISMISSED** [review]
LGTM. Thanks for this!

File: src/sagemaker_training/smdataparallel.py [hunk 1]
```diff
@@ -161,12 +161,15 @@ def _get_mpirun_command(
 
         mpirun_command.extend(additional_options)
 
+        instance_type = self._get_instance_type()
+        # Use EFA's RDMA functionality for one-sided and two-sided transfer
+        if instance_type in ["ml.p3dn.24xlarge", "ml.p4d.24xlarge"]:
+            mpirun_command.extend(["-x", "FI_EFA_USE_DEVICE_RDMA=1"])
```
Comment (indhub) [inline src/sagemaker_training/smdataparallel.py:167]: FI_EFA_USE_DEVICE_RDMA only works on p4d

Comment (karan6181) [reply src/sagemaker_training/smdataparallel.py:167]: Done https://github.com/aws/sagemaker-training-toolkit/pull/101/commits/07cb410433a4ef4cc3a534548c7ae07c83b54198. Thanks!

**Review by piyushghai — APPROVED** [review]

**Review by ChaiBapchya — APPROVED** [review]
Thanks!

**Review by rondogency — APPROVED** [review]

**Review by spidyDev — APPROVED** [review]

**Pull request merged**